secret coordinate helpers

jrhizor opened the pull request:
resolves https://github.com/airbytehq/airbyte/issues/6099

jrhizor commented:
todo:
- [ ] clean up
- [ ] add followup tickets 
- [ ] add docs
- [ ] decide if we want true versions or to always bump (and implement if needed)

airbyte-jenny commented on airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/split_secrets/SecretsHelpersTest.java:
```diff
@@ -0,0 +1,219 @@
+    final var uuidIterator = UUIDS.iterator();
+    final var inputConfig = testCase.getFullConfig();
+    final var inputConfigCopy = inputConfig.deepCopy();
+    final var splitConfig = SecretsHelpers.split(
```
Now that I'm seeing the split method outside of the class it's defined in, I am realizing the name is a bit unclear.  Naming the methods splitConfig and combineConfig might be more obvious.  That way the meaning is apparent even if it's inlined without a matching variable name.

airbyte-jenny commented on airbyte-config/persistence/src/test/java/io/airbyte/config/persistence/split_secrets/SecretsHelpersTest.java:
```diff
@@ -0,0 +1,219 @@
+    assertEquals(testCase.getFirstSecretMap(), splitConfig.getCoordinateToPayload());
+
+    for (Map.Entry<SecretCoordinate, String> entry : splitConfig.getCoordinateToPayload().entrySet()) {
+      System.out.println("entry1 = " + entry);
```
Looks like debug info to remove?  And a bunch more in the following lines.

airbyte-jenny commented on airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/MemorySecretPersistence.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Optional;
+import java.util.function.Function;
+
+public class MemorySecretPersistence implements SecretPersistence {
```
Does this exist for testing purposes only or does it have production use?  Some javadocs would help here.

sherifnada reviewed (commented):
Was mainly looking at test cases, the main one I would suggest adding is a nested oneof (inside an object)

another thing is that it would have been super helpful to have javadocs on the classes and public methods to help the reader understand what's going on.

jrhizor commented:
@sherifnada I fixed the `ArrayOneOfTestCase` and added a `NestedOneOfTestCase`. Is this what you were talking about?

I also cleaned up the `split` function so that should be actually reviewable. 

I'm going to add a ton of docstrings in the morning and slightly improve on some of the naming and then merge if there aren't any objections.

jrhizor commented on airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:
```diff
@@ -0,0 +1,235 @@
+      final var fieldSchemaCombinationType = JsonSecretsProcessor.findJsonCombinationNode(fieldSchema);
+
+      // if the json schema field is an obj and has the airbyte secret field
+      if (fullConfig.has(specProperty)) {
```
I'm a little wary of moving these into private functions because that would hide where the `fullConfig` object is getting updated. I tried restructuring a lot in the other function and outlining what is happening in javadocs.

airbyte-jenny commented on airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretCoordinate.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  public static SecretCoordinate fromFullCoordinate(String fullCoordinate) {
+    final var splits = fullCoordinate.split("_v");
```
Splitting on _v is potentially risky in the future for other secret stores.  A possibly more reliable approach would be to provide a coordinate interface with serialize and deserialize functions that uses its own pojo for each type of secrets vendor.  We don't need that yet, but I wanted to note the risk here.  Maybe leave a comment in the code mentioning that this split may need a revisit.

jrhizor replied:
added a javadoc

airbyte-jenny commented on airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:
```diff
@@ -0,0 +1,421 @@
+   * Used to separate secrets out of a connector configuration and output a partial config that
+   * includes pointers to secrets instead of actual secret values and a map that can be used to update
+   * a {@link SecretPersistence} at coordinates with values from the full config. If the previous
+   * config for this connector's configuration, this method attempts to use the same base coordinate
```
I think a word is missing here.  "If the previous config for this connector's configure [does what?], then..."
